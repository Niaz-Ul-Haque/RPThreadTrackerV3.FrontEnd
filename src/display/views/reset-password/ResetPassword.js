import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { AvForm } from 'availity-reactstrap-validation';
import {
	CardBlock, Button, Row, Col
} from 'reactstrap';
import * as actions from '../../../infrastructure/actions';
import TooltipForm from '../../forms/TooltipForm';
import Card from '../../shared/styled/Card';
import LoadingIndicator from '../../shared/loading/LoadingIndicator';
import ResetPasswordForm from '../../forms/reset-password/ResetPasswordForm';
import { getQuery } from '../../../utility';

const propTypes = {
	displayLoadingIndicator: PropTypes.bool.isRequired,
	resetPasswordError: PropTypes.string,
	submitUserResetPassword: PropTypes.func.isRequired
};
const defaultProps = {
	resetPasswordError: ''
};

const mapStateToProps = (state) => {
	const {
		loading,
		errors
	} = state;
	return {
		displayLoadingIndicator: loading.resetPasswordLoading,
		resetPasswordError: errors.resetPasswordError
	};
};

class ResetPassword extends Component {
	constructor(props) {
		super(props);
		this.handleResetPasswordSubmit = this.handleResetPasswordSubmit.bind(this);
		this.handleInputChange = this.handleInputChange.bind(this);
		this.state = {
			resetPasswordRequest: {
				Email: getQuery().email,
				Code: getQuery().code
			}
		};
	}

	handleResetPasswordSubmit() {
		const { submitUserResetPassword } = this.props;
		const { resetPasswordRequest } = this.state;
		submitUserResetPassword(resetPasswordRequest);
	}

	handleInputChange(event) {
		const { target } = event;
		const { name, value } = target;
		this.setState(prevState => ({
			resetPasswordRequest: Object.assign({}, prevState.resetPasswordRequest, {
				[name]: value
			})
		}));
	}

	render() {
		const { displayLoadingIndicator, resetPasswordError } = this.props;
		let loading = (<span />);
		if (displayLoadingIndicator) {
			loading = (
				<LoadingIndicator style={{
					width: 50,
					height: 50,
					position: 'absolute',
					top: 0,
					right: 0
				}}
				/>
			);
		}
		let error = (<span />);
		if (resetPasswordError) {
			error = (
				<div className="has-danger">
					<p
						data-spec="reset-password-server-error"
						className="form-control-feedback"
					>
						{resetPasswordError}
					</p>
				</div>
			);
		}
		return (
			<Card className="mx-4">
				<CardBlock className="p-4">
					{loading}
					<AvForm
						data-spec="reset-password-form-container"
						onValidSubmit={this.handleResetPasswordSubmit}
					>
						<h1>Reset Password</h1>
						<p className="text-muted">
							Enter your new password below.
						</p>
						{error}
						<TooltipForm
							Renderable={ResetPasswordForm}
							handleInputChange={this.handleInputChange}
						/>
						<Row>
							<Col xs="6">
								<Button color="primary" className="px-4">Request</Button>
							</Col>
							<Col xs="6" className="text-right text-muted">
								<span className="pull-right"><Link href="/login" to="/login">Back to Login</Link></span>

							</Col>
						</Row>
					</AvForm>
				</CardBlock>
			</Card>
		);
	}
}
ResetPassword.propTypes = propTypes;
ResetPassword.defaultProps = defaultProps;
export default connect(mapStateToProps, {
	submitUserResetPassword: actions.submitUserResetPassword
})(ResetPassword);
