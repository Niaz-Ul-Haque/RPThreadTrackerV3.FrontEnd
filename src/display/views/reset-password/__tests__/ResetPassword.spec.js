// #region imports
import React from 'react';
import { getSpecWrapper, shallowWithState } from '../../../../../config/tests/helpers.unit';
import ResetPassword from '../ResetPassword';
// #endregion imports

// #region mocks
jest.mock('../../../shared/loading/LoadingIndicator', () => 'LoadingIndicator');
jest.mock('../../../forms/reset-password/ResetPasswordForm', () => 'ResetPasswordForm');
jest.mock('../../../../infrastructure/actions', () => ({}));
jest.mock('../../../../utility', () => ({
	getQuery: () => ({
		email: 'test@test.com',
		code: '13579'
	})
}));
// #endregion mocks

const createTestProps = propOverrides => ({
	submitUserResetPassword: jest.fn(),
	...propOverrides
});

const createTestState = stateOverrides => ({
	loading: { resetPasswordLoading: false },
	errors: { resetPasswordError: null },
	...stateOverrides
});

describe('rendering', () => {
	describe('snapshots', () => {
		it('should render valid snapshot', () => {
			const props = createTestProps();
			const state = createTestState();
			const jsx = (<ResetPassword {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot with loading indicator when reset password loading', () => {
			const props = createTestProps();
			const state = createTestState({
				loading: {
					resetPasswordLoading: true
				}
			});
			const jsx = (<ResetPassword {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot with error message', () => {
			const props = createTestProps();
			const state = createTestState({ errors: { resetPasswordError: 'An unknown error occurred.' } });
			const jsx = (<ResetPassword {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(element).toMatchSnapshot();
		});
	});
});

describe('behavior', () => {
	describe('handleResetPasswordSubmit', () => {
		it('should be called when form is submitted', () => {
			const submitUserResetPassword = jest.fn();
			const props = createTestProps({ submitUserResetPassword });
			const state = createTestState();
			const jsx = (<ResetPassword {...props} />);
			const element = shallowWithState(jsx, state).dive();
			const form = getSpecWrapper(element, 'reset-password-form-container');
			form.props().onValidSubmit();
			expect(submitUserResetPassword).toHaveBeenCalledTimes(1);
			expect(submitUserResetPassword).toHaveBeenLastCalledWith({ Email: 'test@test.com', Code: '13579' });
		});
	});
	describe('handleInputChange', () => {
		it('should handle text field update', () => {
			const event = { target: { type: 'text', name: 'password', value: 'my-new-password' } };
			const submitUserResetPassword = jest.fn();
			const props = createTestProps({ submitUserResetPassword });
			const state = createTestState();
			const jsx = (<ResetPassword {...props} />);
			const element = shallowWithState(jsx, state).dive();
			element.instance().handleInputChange(event);
			const form = getSpecWrapper(element, 'reset-password-form-container');
			form.props().onValidSubmit();
			expect(submitUserResetPassword).toHaveBeenCalledTimes(1);
			expect(submitUserResetPassword).toHaveBeenLastCalledWith({ Email: 'test@test.com', Code: '13579', password: 'my-new-password' });
		});
	});
});
