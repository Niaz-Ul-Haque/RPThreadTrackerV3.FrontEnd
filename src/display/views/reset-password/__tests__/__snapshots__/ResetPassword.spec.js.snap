// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering snapshots should render valid snapshot 1`] = `
<Styled(Card)
  className="mx-4"
>
  <CardBlock
    className="p-4"
    tag="div"
  >
    <span />
    <AvForm
      data-spec="reset-password-form-container"
      method="get"
      model={Object {}}
      onInvalidSubmit={[Function]}
      onKeyDown={[Function]}
      onSubmit={[Function]}
      onValidSubmit={[Function]}
      tag={[Function]}
      validationEvent={
        Array [
          "onChange",
          "onInput",
        ]
      }
    >
      <h1>
        Reset Password
      </h1>
      <p
        className="text-muted"
      >
        Enter your new password below.
      </p>
      <span />
      <TooltipForm
        Renderable="ResetPasswordForm"
        handleInputChange={[Function]}
      />
      <Row
        tag="div"
      >
        <Col
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
          xs="6"
        >
          <Button
            className="px-4"
            color="primary"
            tag="button"
          >
            Request
          </Button>
        </Col>
        <Col
          className="text-right text-muted"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
          xs="6"
        >
          <span
            className="pull-right"
          >
            <Link
              href="/login"
              replace={false}
              to="/login"
            >
              Back to Login
            </Link>
          </span>
        </Col>
      </Row>
    </AvForm>
  </CardBlock>
</Styled(Card)>
`;

exports[`rendering snapshots should render valid snapshot with error message 1`] = `
<Styled(Card)
  className="mx-4"
>
  <CardBlock
    className="p-4"
    tag="div"
  >
    <span />
    <AvForm
      data-spec="reset-password-form-container"
      method="get"
      model={Object {}}
      onInvalidSubmit={[Function]}
      onKeyDown={[Function]}
      onSubmit={[Function]}
      onValidSubmit={[Function]}
      tag={[Function]}
      validationEvent={
        Array [
          "onChange",
          "onInput",
        ]
      }
    >
      <h1>
        Reset Password
      </h1>
      <p
        className="text-muted"
      >
        Enter your new password below.
      </p>
      <div
        className="has-danger"
      >
        <p
          className="form-control-feedback"
          data-spec="reset-password-server-error"
        >
          An unknown error occurred.
        </p>
      </div>
      <TooltipForm
        Renderable="ResetPasswordForm"
        handleInputChange={[Function]}
      />
      <Row
        tag="div"
      >
        <Col
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
          xs="6"
        >
          <Button
            className="px-4"
            color="primary"
            tag="button"
          >
            Request
          </Button>
        </Col>
        <Col
          className="text-right text-muted"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
          xs="6"
        >
          <span
            className="pull-right"
          >
            <Link
              href="/login"
              replace={false}
              to="/login"
            >
              Back to Login
            </Link>
          </span>
        </Col>
      </Row>
    </AvForm>
  </CardBlock>
</Styled(Card)>
`;

exports[`rendering snapshots should render valid snapshot with loading indicator when reset password loading 1`] = `
<Styled(Card)
  className="mx-4"
>
  <CardBlock
    className="p-4"
    tag="div"
  >
    <LoadingIndicator
      style={
        Object {
          "height": 50,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 50,
        }
      }
    />
    <AvForm
      data-spec="reset-password-form-container"
      method="get"
      model={Object {}}
      onInvalidSubmit={[Function]}
      onKeyDown={[Function]}
      onSubmit={[Function]}
      onValidSubmit={[Function]}
      tag={[Function]}
      validationEvent={
        Array [
          "onChange",
          "onInput",
        ]
      }
    >
      <h1>
        Reset Password
      </h1>
      <p
        className="text-muted"
      >
        Enter your new password below.
      </p>
      <span />
      <TooltipForm
        Renderable="ResetPasswordForm"
        handleInputChange={[Function]}
      />
      <Row
        tag="div"
      >
        <Col
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
          xs="6"
        >
          <Button
            className="px-4"
            color="primary"
            tag="button"
          >
            Request
          </Button>
        </Col>
        <Col
          className="text-right text-muted"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
          xs="6"
        >
          <span
            className="pull-right"
          >
            <Link
              href="/login"
              replace={false}
              to="/login"
            >
              Back to Login
            </Link>
          </span>
        </Col>
      </Row>
    </AvForm>
  </CardBlock>
</Styled(Card)>
`;
